### PR TITLE
Disable a user account

### LINKED ISSUE #1252: Missing disable user registration setting
Like in O1.

### PR DESCRIPTION
**Pull request by Skrypt**
image

## Todo : 
- Add some logic to prevent disabling the only existing admin user account.

Fixes #1252

### CODE REVIEW

**Comment by agriffard** [comment]
Is it possible to have a BulkAction to Enable/Disable users and a way to display if the user is disabled in the SummaryAdmin?

**Comment by Skrypt** [comment]
image

File: src/OrchardCore.Modules/OrchardCore.Users/Views/UserFields.cshtml [hunk 1]
```diff
@@ -1,3 +1,9 @@
 @model SummaryAdminUserViewModel
-    
-<h5>@Model.User.UserName</h5>
\ No newline at end of file
+
+<h5 class="float-left">@Model.User.UserName</h5>
+@if (Model.User.IsDisabled)
+{
+    <span class="badge badge-danger ml-2" title="@T["Disabled"]">
+        <i class="fa fa-check" aria-hidden="true"></i> @T["Disabled"]
```
Comment (agriffard) [inline src/OrchardCore.Modules/OrchardCore.Users/Views/UserFields.cshtml]: https://fontawesome.com/icons/ban or https://fontawesome.com/icons/user-alt-slash could be appropriate.

**Comment by Skrypt** [comment]
image

**Comment by Skrypt** [comment]
There's nothing that I can use to prevent not disabling the first administrator account created on the website. I can't rely on evaluating the User with the lowest Document.Id with SQL. We need to mark a user as a superuser to avoid disabling it.

File: src/OrchardCore.Modules/OrchardCore.Users/Controllers/AdminController.cs [hunk 2]
```diff
@@ -208,7 +209,16 @@ public async Task<ActionResult> IndexPOST(UserIndexOptions options, IEnumerable<
...
+                            await _userManager.UpdateAsync(item);
+                            _notifier.Success(TH["User {0} successfully disabled.", item.UserName]);
+                        }
+                        break;
+                    case UsersBulkAction.Enable:
+                        foreach (var item in checkedContentItems)
+                        {
+                            item.IsDisabled = false;
+                            await _userManager.UpdateAsync(item);
                             _notifier.Success(TH["User {0} successfully disabled.", item.UserName]);
```
Comment (agriffard) [inline src/OrchardCore.Modules/OrchardCore.Users/Controllers/AdminController.cs]: enabled

**Comment by hishamco** [comment]
> There's nothing that I can use to prevent not disabling the first administrator account created on the website.

For now we can avoid disabling the first created user, but we need to have a role for that or at least name it `superadmin` or `superuser` by default

**Comment by sebastienros** [comment]
Looks weird that the flag is green when the user is disabled

**Comment by hishamco** [comment]
Seb if you mean the alert it makes sense that the operation is successfully done

**Comment by Skrypt** [comment]
Yeah, it's maybe confusing for some but I mean we are consistent throughout the entire UI with those so I'm not sure 100%.

**Comment by sebastienros** [comment]
No, the switch. The ui should be the opposite (keep the bool as `disabled` though)

**Comment by Skrypt** [comment]
Ah! yeah let me change that color. Must be an easy fix.

**Comment by hishamco** [comment]
> No, the switch. The ui should be the opposite (keep the bool as disabled though)

I'm with you on this one :)

**Comment by hishamco** [comment]
Or you can make the default status "Enable", coz the off in the UI will be grayed, which indicate it's disabled

**Comment by Skrypt** [comment]
IsDisabled = false by default as this param was not even there before so it will work even if the value is not present in db.

**Comment by hishamco** [comment]
I see, but almost the apps, CMSs enable the account when it's created. I know it's useful for backward compatibility, but while we will release 1.0.0, such change has not a big impact

**Comment by Skrypt** [comment]
image

**Comment by Skrypt** [comment]
image

**Comment by Skrypt** [comment]
I think this can be merged. We can still enable a user from DB if someone makes a mistake. Else @sebastienros give me some guidance on how you would want this to be handled. We need to have a way to avoid to disable all users and to always keep at least 1 admin user enabled.

**Comment by Skrypt** [comment]
Fixed merge conflicts

**Comment by sebastienros** [comment]
REbuild assets.

Can the switch text and input be changed to Enabled? I don't like switches that express negativity.

**Comment by Skrypt** [comment]
Yes, but then it will require to change the name to IsEnabled and create a migration for existing users.
Then change the default value of that db colum to be true by default. Need to see if it will set all users to enabled just by adding the column with a default of 1.

**Comment by Skrypt** [comment]
Works if the column is non-null.

**Comment by sebastienros** [comment]
No, I am just talking about the UI

**Comment by Skrypt** [comment]
Ok I changed it to IsEnabled. I've been probably too quick on doing the change.

**Comment by Skrypt** [comment]
Ok I made a test. In the Document table I removed the "IsEnabled: true" JSON data in the Content column for a user that existed. In the UserIndex table the IsEnabled = 1 still so when editing a User it still knows that it's enabled.

So this should work fine with the migration since it will set a value of IsEnabled = 1 by default to all Users in the UserIndex table when executed.

Also tested logging in with that same User and it worked.

**Pull request merged**